hono/src/services/items.ts

```diff
@@ -23,8 +23,54 @@ const selectItemSchema = createSelectSchema(items, {
   id: z.string().refine((val) => isCuid(val)),
 })
 
-app.get('/', injectDB, async (c) => {
-  return c.json(await c.get('db').query.items.findMany())
+const paginationValidation = z.object({
+  page: z
+    .string()
+    .refine((val) => !isNaN(+val) && +val >= 0, {
+      message: 'Page number must be a non-negative number',
+    })
+    .optional(),
+  size: z
+    .string()
+    .refine((val) => !isNaN(+val) && +val > 0, {
+      message: 'Page size must be a positive number',
+    })
+    .optional(),
+})
+
+app.get('/', zValidator('query', paginationValidation), injectDB, async (c) => {
+  const { page, size } = c.req.query()
+
+  const pageNumber: number = page ? +page : 0
+  const pageSize: number = size ? +size : 25
+
+  const itemsData = await c
+    .get('db')
+    .select()
+    .from(items)
+    .where(eq(items.authorUsername, 'jdoe'))
+    .limit(pageSize)
+    .offset(pageNumber)
+
+  await c.get('db').execute(sql`
+      ANALYZE items;
+    `)
+
+  const estimate = await c.get('db').execute(sql`
+      SELECT reltuples AS estimate
```

Comment:
Tried using drizzle's built-in functions. This is the alternative.

Reply:
This looks appropriate. What guarantees (or lack thereof) does this query give in terms of estimating? How often does it update the total count?

Reply:
I've locally tested this query to see how it would update in response to adding more items (I did have to explicitly do `ANALYZE ITEMS` for each request to see the total updated locally). This total count is updated every time the database vacuums or analyzes. I am not exactly sure if postgres is set up to auto vacuum however.

Source [used](https://wiki.postgresql.org/wiki/Count_estimate).

Reply:
It does appear that Neon (our production Postgres instance) does have auto vacuum configured https://community.neon.tech/t/does-neon-run-autovacuum/1134. I'd like to play around with the staging env and see how long it takes before the total is updated (mins? hours? days?). Ideally something on the order of 10 mins seems fair.

Reply:
It has been around 14 hours and the total seems to have stayed the same despite me adding more items locally. Perhaps there is a configuration that we are missing?

Reply:
The best fix here seems to run `ANALYZE items;` as a separate, non-blocking query to API calls where items are either created or deleted (POST, DELETE). The most important part here is to make sure that execution is not blocked or waiting on ANALYZE to finish.

Reply:
No longer relevant, closing.